provider/ultradns/ultradns.go

```diff
@@ -0,0 +1,502 @@
+
+	log "github.com/sirupsen/logrus"
+	udnssdk "github.com/ultradns/ultradns-sdk-go"
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+	"sigs.k8s.io/external-dns/provider"
+)
+
+const (
+	ultradnsCreate = "CREATE"
+	ultradnsDelete = "DELETE"
+	ultradnsUpdate = "UPDATE"
+	sbPoolPriority = 1
+	sbPoolOrder    = "ROUND_ROBIN"
+	rdPoolOrder    = "ROUND_ROBIN"
+)
+
+// global variables
+var sbPoolRunProbes = true
+var sbPoolActOnProbes = true
+var ultradnsPoolType = "rdpool"
+
+//Setting custom headers for ultradns api calls
+var customHeader = []udnssdk.CustomHeader{
+	{
+		Key:   "UltraClient",
+		Value: "kube-client",
+	},
+}
+
+// UltraDNSProvider struct
+type UltraDNSProvider struct {
+	provider.BaseProvider
+	client       udnssdk.Client
+	domainFilter endpoint.DomainFilter
+	DryRun       bool
+	AccountName  string
+}
+
+// UltraDNSChanges struct
+type UltraDNSChanges struct {
+	Action string
+
+	ResourceRecordSetUltraDNS udnssdk.RRSet
+}
+
+// NewUltraDNSProvider initializes a new UltraDNS DNS based provider
+func NewUltraDNSProvider(domainFilter endpoint.DomainFilter, dryRun bool) (*UltraDNSProvider, error) {
+	Username, ok := os.LookupEnv("ULTRADNS_USERNAME")
+	udnssdk.SetCustomHeader = customHeader
+	if !ok {
+		return nil, fmt.Errorf("no username found")
+	}
+
+	Base64Password, ok := os.LookupEnv("ULTRADNS_PASSWORD")
+	if !ok {
+		return nil, fmt.Errorf("no password found")
+	}
+
+	// Base64 Standard Decoding
+	Password, err := base64.StdEncoding.DecodeString(Base64Password)
+	if err != nil {
+		fmt.Printf("Error decoding string: %s ", err.Error())
+		return nil, err
+	}
+
+	BaseURL, ok := os.LookupEnv("ULTRADNS_BASEURL")
+	if !ok {
+		return nil, fmt.Errorf("no baseurl found")
+	}
+	AccountName, ok := os.LookupEnv("ULTRADNS_ACCOUNTNAME")
+	if !ok {
+		AccountName = ""
+	}
+
+	probeValue, ok := os.LookupEnv("ULTRADNS_ENABLE_PROBING")
+	if ok {
+		if (probeValue != "true") && (probeValue != "false") {
+			return nil, fmt.Errorf("please set proper probe value, the values can be either true or false")
+		}
+		sbPoolRunProbes, _ = strconv.ParseBool(probeValue)
+	}
+
+	actOnProbeValue, ok := os.LookupEnv("ULTRADNS_ENABLE_ACTONPROBE")
+	if ok {
+		if (actOnProbeValue != "true") && (actOnProbeValue != "false") {
+			return nil, fmt.Errorf("please set proper act on probe value, the values can be either true or false")
+		}
+		sbPoolActOnProbes, _ = strconv.ParseBool(actOnProbeValue)
+	}
+
+	poolValue, ok := os.LookupEnv("ULTRADNS_POOL_TYPE")
+	if ok {
+		if (poolValue != "sbpool") && (poolValue != "rdpool") {
+			return nil, fmt.Errorf(" please set proper ULTRADNS_POOL_TYPE, supported types are sbpool or rdpool")
+		}
+		ultradnsPoolType = poolValue
+	}
+
+	client, err := udnssdk.NewClient(Username, string(Password), BaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("connection cannot be established")
+	}
+
+	provider := &UltraDNSProvider{
+		client:       *client,
+		domainFilter: domainFilter,
+		DryRun:       dryRun,
```

Comment:
Please don't mix. Your fields should stay private, you only need it in your package

Reply:
Made appropriate changes